### DISCUSSION TITLE
Add support for -Yscala-release flag

### DISCUSSION THREAD

**Discussion by prolativ** [sbt/sbt#6812] [Ideas]
Scala 3.1.2-RC1 introduces a new experimental compiler flag which allows scalac to produce compilation outputs consumable for older language minor versions https://github.com/lampepfl/dotty/pull/14156.
E.g. a library scala project compiled using the 3.1.2-RC1 compiler with `-Yscala-release 3.0` flag can be used as a dependency for some other project compiled with scala 3.0.2.

This works smoothly in simple cases when the classpath of a project is managed by hand and we use locally produced artifacts but this breaks if the library was published and it's added to the project as a managed dependency. The root cause of the problem is that projects compiled with scala 3.1.2-RC1 automatically get 3.1.2-RC1 standard library as their dependency, which ends up in the downstream project's classpath making an older 3.0.x compiler crash because it cannot read TASTy files in the newer format.

The currently known workaround for this is to add dependency overrides in the downstream project:
```scala
dependencyOverrides ++= Seq(
  scalaOrganization.value %% "scala3-library" % scalaVersion.value,
  scalaOrganization.value %% "scala3-library_sjs1" % scalaVersion.value // for Scala.js projects
)
```

However this is imperfect as users of a library shouldn't have to know whether it was compiled with a 3.0.x compiler or with 3.1.y with `-Yscala-release 3.0` flag. Adding these overrides by default wouldn't solve the problem entirely.

It seems to me that the right approach would be to change the way how pom.xml and ivy.xml files are generated during publishing a project so that they add the older standard library (matching the argument of `-Yscala-release`) to the list of dependencies instead of the one provided by the compiler used to compile the project.

I tried to implement this by myself but I didn't find an elegant way to separate the dependencies used to compile a project from the ones declared as dependencies in pom.xml and ivy.xml. My guess would be that I have to split `allDependencies` into two separate task keys used in these two contexts and I'm wondering if this can be done nicely in sbt 1.x without breaking backward compatibility.

It's also not entirely clear which version of the standard library should be used as the dependency when a project is compiled with `-Yscala-release 3.0`. Some possibilities to consider are:
* use the oldest version from each minor release series (3.x.0) hoping that it will be bumped to a newer version by eviction when added as a dependency - easiest to implement but the version might not get bumped when the library is used as a dependency for a java project or when it's used as a standalone project rather than a library
* use the newest available version from each minor release series (resolved when the library project is compiled) - we'll get all available fixes for the stdlib but this is somehow not fully deterministic as the same codebase compiled twice with some time between the compilations would add different dependencies to pom.xml and ivy.xml
* use the newest version from each minor release series available when the compiler compiling the project gets released - that might require storing a mapping somewhere: it would not be very convenient to hardcode it in sbt's codebase so probably then `scalac` could have a command mapping a release version to a specific version of stdlib (including a patch number)

As `-Yscala-release 3.1` has no effect for a `3.1.x` compiler the stdlib version in pom.xml and ivy.xml should not change either in such a case, which seems to align best with the last option.

We also need to keep in mind that this will somehow need to be handled for scala.js (and probably for scala-native although this flag hasn't been tested with it yet) so this issue might relate to https://github.com/sbt/sbt/discussions/6736.

Perfectly we would like to make `-Yscala-release` work in sbt out of the box before the release of scala 3.2.0.

**Comment by eed3si9n** [comment]
Let's say using `-Yscala-release 3.0` with 3.1.2-RC1, we're building `foo-json`:
- Does the compiler expect scala3-library 3.1.2-RC1 at compile-time of `foo-json`, or it just needs scala3-library 3.0.x?
- It sounds like the downstream user `bar-app` require only scala3-library 3.0 at its compile-time?
- And I am assuming that `bar-app` requires only scala3-library 3.0 at runtime.

**Reply by prolativ** [reply]
Exactly, `foo-json` needs to have `3.1.2-RC1` stdlib on it's classpath during its compilation but after that all projects using `foo-json` as a dependency can use any stdlib in version 3.0 or 3.1 (or later in the future) at compile-time and at runtime (of course assuming that the runtime version has to be greater than or equal to the compile-time version but from the perspective of pom.xml and ivy.xml files they will be typically the same)

**Comment by prolativ** [comment]
Here's my PR implementing this https://github.com/sbt/sbt/pull/6814/files
